cms/djangoapps/contentstore/tasks.py

```diff
@@ -808,3 +813,159 @@ def handle_course_import_exception(courselike_key, exception, status, known=True
+        return 'None'
+    return Organization.objects.get(short_name=org_name)
+
+
+def copy_v1_user_roles_into_v2_library(v2_library_key, v1_library_key):
+    """
+    write the access and edit permissions of a v1 library into a v2 library.
+    """
+
+    def _get_users_by_access_level(v1_library_key):
+        """
+        Get a permissions object for a library which contains a list of user IDs for every V2 permissions level,
+        based on V1 library roles.
+        The following mapping exists for a library:
+        V1 Library Role -> V2 Permission Level
+        LibraryUserRole -> READ_LEVEL
+        CourseStaffRole -> AUTHOR_LEVEL
+        CourseInstructorRole -> ADMIN_LEVEL
+        """
+        permissions = {}
+        permissions[v2contentlib_api.AccessLevel.READ_LEVEL] = list(LibraryUserRole(v1_library_key).users_with_role())
+        permissions[v2contentlib_api.AccessLevel.AUTHOR_LEVEL] = list(CourseStaffRole(v1_library_key).users_with_role())
+        permissions[v2contentlib_api.AccessLevel.ADMIN_LEVEL] = list(
+            CourseInstructorRole(v1_library_key).users_with_role()
+        )
+        return permissions
+
+    permissions = _get_users_by_access_level(v1_library_key)
+    for access_level in permissions.keys():  # lint-amnesty, pylint: disable=consider-iterating-dictionary
+        for user in permissions[access_level]:
+            v2contentlib_api.set_library_user_permissions(v2_library_key, user, access_level)
+
+
+def _create_copy_content_task(v2_library_key, v1_library_key):
+    """
+    spin up a celery task to import the V1 Library's content into the V2 library.
+    This utalizes the fact that course and v1 library content is stored almost identically.
+    """
+    return v2contentlib_api.import_blocks_create_task(v2_library_key, v1_library_key)
+
+
+def _create_metadata(v1_library_key, collection_uuid):
+    """instansiate an index for the V2 lib in the collection"""
+
+    store = modulestore()
+    v1_library = store.get_library(v1_library_key)
+    collection = get_collection(collection_uuid).uuid
+    # To make it easy, all converted libs are complex, meaning they can contain problems, videos, and text
+    library_type = 'complex'
+    org = _parse_organization(v1_library.location.library_key.org)
+    slug = v1_library.location.library_key.library
+    title = v1_library.display_name
+    #  V1 libraries do not have descriptions.
+    description = ''
+    #  permssions & license are most restrictive.
+    allow_public_learning = False
+    allow_public_read = False
+    library_license = ''  # '' = ALL_RIGHTS_RESERVED
+    with atomic():
+        return v2contentlib_api.create_library(
+            collection,
+            library_type,
+            org,
+            slug,
+            title,
+            description,
+            allow_public_learning,
+            allow_public_read,
+            library_license
+        )
+
+
+@shared_task(time_limit=30)
```

Comment:
I presume the 30 minute limit is based on some empirical data. Do we know what a "typical" library copy takes to do? Do we know the dynamic range in library size? What happens on a timeout?

Reply:
> 30 minute limit

30 seconds! 

> Do we know what a "typical" library copy takes to do

The average block has 3 users with permissions, and 6 blocks. It takes about 7 seconds to "copy." A large amount of that time is spinning up a management command.

> Do we know the dynamic range in library size?

The largest v1 library has 104 blocks. That worst-case runtime was 25 seconds.

> What happens on a timeout?

The task returns as an error, and that error is logged.